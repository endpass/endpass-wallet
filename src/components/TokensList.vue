<template>
  <div class="tokens-list" data-test="tokens-list">
    <ul v-if="tokens.length > 0">
      <li
        v-for="token in tokens"
        :class="itemClass"
        :key="token.address"
        data-test="user-token"
      >
        <v-token :token="token" :currency="currency">
          <a
            v-if="isTokenCanBeDeleted(token)"
            slot="right"
            class="is-inline-block remove-token-button"
            title="Remove Token"
            data-test="delete-button"
            @click="deleteToken(token)"
          >
            <span
              class="icon has-text-danger is-small is-pulled-right"
              v-html="require('@/img/x.svg')"
            />
          </a>
        </v-token>
      </li>
    </ul>
    <p v-else class="small">
      You have no tokens at this address.
    </p>
  </div>
</template>

<script>
import get from 'lodash/get';
import VToken from '@/components/VToken';
import { mapState, mapActions, mapGetters } from 'vuex';
import error from '@/mixins/error';

export default {
  props: {
    tokens: {
      type: Array,
      default: () => [],
    },

    hasRemove: {
      type: Boolean,
      default: false,
    },

    itemClass: {
      type: [Object, Array, String],
      default: '',
    },
  },

  computed: {
    ...mapGetters('tokens', ['currentNetUserFullTokens']),
    ...mapState({
      currency: state => state.user.settings.fiatCurrency,
      userTokens: state => state.tokens.userTokens,
    }),
  },

  methods: {
    ...mapActions('tokens', ['removeUserToken']),

    isTokenCanBeDeleted(token) {
      const { hasRemove, currentNetUserFullTokens } = this;

      return (
        hasRemove &&
        currentNetUserFullTokens[token.address] &&
        get(token, 'balance', '0') === '0'
      );
    },

    async deleteToken(token) {
      await this.removeUserToken({
        token,
      });
    },
  },

  mixins: [error],

  components: {
    VToken,
  },
};
</script>

<style lang="scss"></style>
