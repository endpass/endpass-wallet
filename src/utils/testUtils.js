import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
