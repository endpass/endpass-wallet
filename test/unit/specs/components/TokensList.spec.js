import { createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import { wrapShallowMountFactory, wrapMountFactory } from '@/testUtils';

import TokensList from '@/components/TokensList';

import {
  zeroBalancedTokens,
  dustAmountTokens,
  valuableTokens,
  tokensWithBalancesMappedByAddresses,
} from 'fixtures/tokens';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('TokensList', () => {
  let actions;
  let wrapper;
  let wrapperFactory;
  let store;
  let options;

  beforeEach(() => {
    actions = {
      getTokensPrices: jest.fn(),
      removeUserToken: jest.fn(),
    };

    store = new Vuex.Store({
      modules: {
        user: {
          namespaced: true,
          state: {
            settings: {
              fiatCurrency: 'USD',
            },
          },
        },
        tokens: {
          namespaced: true,
          state: {
            prices: {
              FST: { ETH: 2 }, // price of token in ETH
            },
          },
          getters: {
            currentNetUserFullTokens: () => tokensWithBalancesMappedByAddresses,
          },
          actions,
        },
        price: {
          namespaced: true,
          state: {
            price: 1, // Ethereum price
          },
        },
        errors: {
          namespaced: true,
          actions: {
            emitError(state, e) {
              throw e;
            },
          },
        },
      },
    });

    options = {
      store,
      localVue,
      propsData: {
        tokens: valuableTokens,
      },
    };
    wrapperFactory = wrapShallowMountFactory(TokensList, options);
    wrapper = wrapperFactory();
  });

  describe('render', () => {
    it('should render given tokens list', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should not show remove token button by default', () => {
      expect(wrapper.find('.remove-token-button').exists()).toBeFalsy();
    });

    it('should not render dust amount tokens by default', () => {
      wrapper = wrapperFactory({
        store,
        localVue,
        propsData: {
          tokens: dustAmountTokens,
        },
      });

      expect(wrapper.find('[data-test=dustbin-toggler]').exists()).toBe(true);
      expect(wrapper.find('[data-test=user-token]').exists()).toBe(false);
    });

    it('should render dust amount tokens if toggler is enabled', () => {
      wrapper = wrapperFactory({
        store,
        localVue,
        propsData: {
          tokens: dustAmountTokens,
        },
      });

      expect(wrapper.find('[data-test=user-token]').exists()).toBe(false);

      wrapper.setData({
        isDustbinTokensVisible: true,
      });

      expect(wrapper.find('[data-test=user-token]').exists()).toBe(true);
    });

    it('should render valuable tokens by default', () => {
      wrapper = wrapperFactory({
        store,
        localVue,
        propsData: {
          tokens: [...dustAmountTokens, ...valuableTokens],
        },
      });

      expect(wrapper.find('[data-test=user-token]').exists()).toBe(true);
    });

    it('should not render dustbin controls if collapsable is falsy', () => {
      wrapper = wrapperFactory({
        store,
        localVue,
        propsData: {
          tokens: dustAmountTokens,
          collapsable: false,
        },
      });

      expect(wrapper.find('[data-test=dustbin-toggler]').exists()).toBe(false);
      expect(wrapper.find('[data-test=dustbin-user-token]').exists()).toBe(
        false,
      );
      expect(wrapper.find('[data-test=user-token]').exists()).toBe(true);
    });
  });

  describe('behavior', () => {
    beforeEach(() => {
      options = {
        ...options,
        propsData: {
          hasRemove: true,
          tokens: zeroBalancedTokens,
        },
        computed: {
          currentNetUserFullTokens: tokensWithBalancesMappedByAddresses,
        },
      };
      wrapperFactory = wrapMountFactory(TokensList, options);
    });

    describe('remove token', () => {
      beforeEach(() => {
        wrapper = wrapperFactory({
          ...options,
          methods: {
            isTokenCanBeDeleted: jest.fn().mockImplementationOnce(() => true),
          },
        });
        wrapper.setData({
          isDustbinTokensVisible: true,
        });
      });

      it('should render remove button only for user tokens with zero balance', () => {
        expect(wrapper.find('[data-test=delete-button]').exists()).toBe(true);
      });

      it('should remove tokens on click remove button', () => {
        jest.spyOn(wrapper.vm, 'deleteToken');

        wrapper.find('[data-test=delete-button]').trigger('click');

        expect(wrapper.vm.deleteToken).toBeCalledTimes(1);
      });
    });

    describe('dustbin tokens tokens', () => {
      it('should show dustbin tokens on toggler check', () => {
        wrapper = wrapperFactory({
          store,
          localVue,
          propsData: {
            tokens: dustAmountTokens,
          },
        });
        wrapper.setData({
          isDustbinTokensVisible: false,
        });

        expect(wrapper.find('[data-test=user-token]').exists()).toBe(false);

        wrapper.find('[data-test=dustbin-toggler]').trigger('click');

        expect(wrapper.vm.isDustbinTokensVisible).toBe(true);
        expect(wrapper.find('[data-test=user-token]').exists()).toBe(true);
      });
    });
  });
});
