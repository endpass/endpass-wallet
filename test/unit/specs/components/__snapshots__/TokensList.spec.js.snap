// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokensList render should render given tokens list 1`] = `
<div data-test="tokens-list" class="tokens-list">
    <ul>
        <li data-test="user-token" class="">
            <v-token-stub token="[object Object]" currency="USD">
                <!---->
            </v-token-stub>
        </li>
        <li data-test="user-token" class="">
            <v-token-stub token="[object Object]" currency="USD">
                <!---->
            </v-token-stub>
        </li>
    </ul>
</div>
`;
